### PR TITLE
Preprocessing: faster build vocab + multiple weighted datasets

### PR DESCRIPTION
**Pull request by francoishernandez**
The current preprocessing works in the following manner:
- build train (from a single src and tgt pair of files) and valid shards;
- build vocabulary from train shards.

This means train shards are re-loaded after having been dumped.
We have in mind to simplify this by creating the vocabulary along with the train shards.

For now I've just taken the logic from `onmt.inputters.inputter.build_vocab` and inserted the necessary bits in `preprocess.build_save_dataset`.
It might not be the cleanest way to do, but it's a start.

Tested on toy text / speech / image datasets, and it seems to work fine.

Glad to have some feedback on a cleaner way to refactor preprocessing codepath(s) @guillaumekln @flauted @bpopeters .

Next ideas would be, either in this PR or in a following one:
- [x] allow several corpora with different weights (upsampling);
- [ ] allow multiple threads while creating shards.

PS: @flauted while testing stuff for this PR, I stumbled upon a maybe unwanted behaviour introduced [here](https://github.com/OpenNMT/OpenNMT-py/pull/1216/files#diff-55eec480db97bda4ea4ea8fc498bfad6R260). This, in combination with some `-src_words_min_frequency` or `tgt_words_min_frequency` option may lead to removing some of the last tokens of the vocab.

### CODE REVIEW

**Comment by flauted** [comment]
About the PS: Good catch. It's older than that though, see [here](https://github.com/OpenNMT/OpenNMT-py/pull/1216/files#diff-55eec480db97bda4ea4ea8fc498bfad6L284). I haven't really used the text file vocab feature, but I'm pretty sure it's just a list of (theoretically unique) words. So the counts/freqs are lost. The validation should probably check for non-default min freq/count type options if there's a source vocab path. I don't know if that's in the scope of this PR or should be done separately, though.

**Comment by francoishernandez** [comment]
I'm extending the scope of this PR, adding the possibility to pass several train corpora when preprocessing, and give them specific weights when training.

### Preprocessing
I introduce `-train_ids` which is a list of IDs that will be given to the preprocessed shards.
E.g. we have two corpora : `parallel.en` and  `parallel.de` + `from_backtranslation.en` `from_backtranslation.de`, we can pass 
```
...
-train_src parallel.en from_backtranslation.en \
-train_tgt parallel.de from_backtranslation.de \
-train_ids A B \
-save_data my_data \
...
```
and it will dump `my_data.train_A.X.pt` based on `parallel.en`//`parallel.de` and `my_data.train_B.X.pt` based on `from_backtranslation.en`//`from_backtranslation.de`.

### Training
I introduce `-data_ids` based on the same principle as above, as well as `-data_weights`, which is the list of the weight each corpus should have.
E.g.
```
...
-data my_data \
-data_ids A B \
-data_weights 1 7 \
...
```
will mean that we'll look for `my_data.train_A.*.pt` and `my_data.train_B.*.pt`, and that when building batches, we'll take 1 example from corpus A, then 7 examples from corpus B, and so on.

For this purpose, I created the `MultipleDatasetIterator` class in `onmt.inputters.inputter` which takes a list of `DatasetLazyIter` and the respective list of weights to yield the batch according to the rule described above.

File: onmt/inputters/inputter.py [hunk 1]
```diff
@@ -635,9 +745,15 @@ def build_dataset_iter(corpus_type, fields, opt, is_train=True):
...
-    batch_fn = max_tok_len if is_train and opt.batch_type == "tokens" else None
-    batch_size_multiple = 8 if opt.model_dtype == "fp16" else 1
+    if multi:
+        batch_size = 1
+        batch_fn = None
+        batch_size_multiple = 1
+    else:
+        batch_size = opt.batch_size if is_train else opt.valid_batch_size
+        batch_fn = max_tok_len \
+            if is_train and opt.batch_type == "tokens"else None
```
Comment (vince62s) [inline onmt/inputters/inputter.py]: typo missing space before else

File: onmt/inputters/inputter.py [hunk 2]
```diff
@@ -635,9 +745,15 @@ def build_dataset_iter(corpus_type, fields, opt, is_train=True):
         glob.glob(opt.data + '.' + corpus_type + '*.pt')))
     if not dataset_paths:
         return None
-    batch_size = opt.batch_size if is_train else opt.valid_batch_size
-    batch_fn = max_tok_len if is_train and opt.batch_type == "tokens" else None
-    batch_size_multiple = 8 if opt.model_dtype == "fp16" else 1
+    if multi:
+        batch_size = 1
+        batch_fn = None
+        batch_size_multiple = 1
```
Comment (vince62s) [inline onmt/inputters/inputter.py]: might be the reason why we lose perf in FP16. we need a multiple of 8

File: preprocess.py [hunk 3]
```diff
@@ -32,51 +35,111 @@ def build_save_dataset(corpus_type, fields, src_reader, tgt_reader, opt):
...
+        else:
+            filter_pred = None
+
+        if corpus_type == "train":
+            if opt.src_vocab:
+                try:
+                    logger.info("Using existing vocabulary...")
+                    vocab = torch.load(opt.src_vocab)
+                    # return vocab to dump with standard name
+                    return vocab
```
Comment (vince62s) [inline preprocess.py]: this does not work. we need src_vocab = torch.load ....
and no return.

File: preprocess.py [hunk 4]
```diff
@@ -32,51 +35,111 @@ def build_save_dataset(corpus_type, fields, src_reader, tgt_reader, opt):
...
+                except torch.serialization.pickle.UnpicklingError:
+                    logger.info("Building vocab from text file...")
+
+            # Load vocabulary
+            if opt.src_vocab:
```
Comment (vince62s) [inline preprocess.py]: this need to indented under the except above

File: preprocess.py [hunk 5]
```diff
@@ -32,51 +35,111 @@ def build_save_dataset(corpus_type, fields, src_reader, tgt_reader, opt):
...
+            gc.collect()
+
+    if corpus_type == "train":
+        fields = _build_fields_vocab(
+            fields, counters, opt.data_type,
+            opt.share_vocab, opt.vocab_size_multiple,
+            opt.src_vocab_size, opt.src_words_min_frequency,
+            opt.tgt_vocab_size, opt.tgt_words_min_frequency)
+        vocab_path = opt.save_data + '.vocab.pt'
+        torch.save(fields, vocab_path)
```
Comment (vince62s) [inline preprocess.py]: even if there is an existing vocab.pt it will be overwritten by this.

**Comment by vince62s** [comment]
beyond the few comments, it does not work. I think the vocab built this way is wrong.
tried a quick training and does not work properly.

**Comment by vince62s** [comment]
Main issue was my fault, seems to work fine now.
Will merge.

**Pull request merged**